api/registration_handler.py

```diff
@@ -0,0 +1,116 @@
+        new_user = ""
+        with DRIVER.session() as session:
+            user_exists = session.write_transaction(
+                get_user, email)
+            if not user_exists:
+                utf8_pw = password.encode('utf-8')
+                hashed_pw = hashpw(utf8_pw, gensalt()).decode('utf-8')
+                new_user = session.write_transaction(
+                    create_user, email, hashed_pw)
+        if new_user != "":
+            session_token = jwt.encode({'email': email},
+                                       self.settings['jwt_secret'],
+                                       algorithm='HS256')
+            self.set_secure_cookie("user", session_token)
+            self.write(json.dumps({
+                "email": email,
+                "session_token": session_token.decode('utf-8')
+            }))
+        else:
+            self.set_status(400)
+            self.finish(json.dumps({
+                'error': {
+                    'code': 400,
+                    'message': "A user with that email already exists!",
+                }
+            }))
+
+    def get(self):
+        """Gets exising user and signs them in"""
+        email = self.get_query_argument('email')
+        password = self.get_query_argument('password')
+        error_message = ""
+        with DRIVER.session() as session:
+            user = session.write_transaction(
+                get_user, email)
+            hashed_pw = user[0].get("password")
+
+        hashedpw_utf8 = hashed_pw.encode('utf-8')
+        pw_utf8 = password.encode('utf-8')
+        if not user:
+            error_message = """No user with that email and password
+                                combination exists!"""
+        elif hashpw(pw_utf8, hashedpw_utf8) == hashedpw_utf8:
+            # do we check if a cookie exists? If so, what do we do? Overwrite?
+            session_token = jwt.encode({'email': email},
+                                       self.settings['jwt_secret'],
+                                       algorithm='HS256')
+            self.set_secure_cookie("user", session_token)
+            return self.finish(json.dumps({
+                "email": user[0].get("email"),
+                "session_token": session_token.decode('utf-8')
+            }))
+        else:
+            error_message = "Incorrect password"
+
+        self.set_status(400)
+        return self.finish(json.dumps({
+            'error': {
+                'code': 400,
+                'message': error_message,
+            }
+        }))
+
+    def options(self):
+        # no body
+        self.set_status(204)
+        self.finish()
+
+
+def get_user(txn, email):
+    """Gets user node."""
+    query = """
+    MATCH (a:User)
+    WHERE a.email = "{email}"
+    RETURN a
+    """.format(email=email)
+    result = txn.run(query)
+    return result.value()
+
+
+def create_user(txn, email, password):
+    """Creates a user node."""
+    query = """
+    MERGE (a:User {email: $email,
```

Comment:
We probably need a bit more validation going on before using this. Right now this would make a new user if you supplied a username that currently exists but a password that is different from the existing user's. I think you guard against this in the Python code, but it's probably best to guard again in the query.

1. Run a match query for the email. If it exists, we should raise an error.
1. If the match doesn't return anything, go ahead and do the merge.